### PR TITLE
Implement charging for Ethereum transactions

### LINKED ISSUE #3221: Implement Gas Charges as per new economics section in HIP-410

### PR DESCRIPTION
**Pull request by dimitar-dinev**
**Description**:

**Related issue(s)**:

Fixes #3221 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

**Comment by shemnon** [comment]
The logic didn't handle the case of a partial sender payment covered by a relay allowance.  Two cases in particular, where the sender amount plus the allowance amount is insufficient to execute the TX should fail pre-execution.  The other case where the sender is a partial amount and the allowance is enough to cover the balance results in a full charge the sender.

I created a matrixed test to test all 9 cases (sender sends zero, partial, full payment and relayer offers zero, partial, and fill coverage) and pushed it.

File: test-clients/src/main/java/com/hedera/services/bdd/suites/ethereum/EthereumSuite.java [hunk 1]
```diff
@@ -52,26 +48,189 @@ public class EthereumSuite extends HapiApiSuite {
...
+				new Object[] { false, partialFee, partialPayment, false, false },
+//				new Object[] { true, partialFee, fullAllowance, false, false }, // done in bothPayFeesSucceeds()
+				new Object[] { true, gasPrice, noPayment, true, false },
+				new Object[] { true, gasPrice, partialPayment, true, false },
+				new Object[] { true, gasPrice, fullAllowance, true, false}
+		).map(params ->
+				// [0] - success
+				// [1] - sender gas price
+				// [2] - relayer offered
+				// [3] - is sender charged
```
Comment (shemnon) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/ethereum/EthereumSuite.java:143]: _if_ they are charged is insufficient.  (a) we need to validate _how much_ especially in the mixed payment scenario and (b) this isn't asserting that zero fees are charged when set to zero

File: hedera-node/src/main/java/com/hedera/services/txns/contract/ContractCallTransitionLogic.java [hunk 2]
```diff
@@ -119,18 +125,32 @@ public void doStateTransitionOperation(final TransactionBody contractCallTxn, fi
 				: Bytes.EMPTY;
 
 		// --- Do the business logic ---
-		if (incrementCounter) {
+		if (relayerId != null) {
```
Comment (Nana-EC) [inline hedera-node/src/main/java/com/hedera/services/txns/contract/ContractCallTransitionLogic.java:128]: nit: you could move the contents of this if to the `TransactionProcessingResult` assigning else below for simplicity

File: hedera-node/src/main/java/com/hedera/services/txns/contract/ContractCreateTransitionLogic.java [hunk 3]
```diff
@@ -151,7 +156,7 @@ public void doStateTransitionOperation(final TransactionBody contractCreateTxn,
 
 
 		// --- Do the business logic ---
-		if (incrementCounter) {
+		if (relayerId != null) {
```
Comment (Nana-EC) [inline hedera-node/src/main/java/com/hedera/services/txns/contract/ContractCreateTransitionLogic.java:159]: Same comment here, could move it down

File: hedera-node/src/test/java/com/hedera/services/contracts/execution/CallEvmTxProcessorTest.java [hunk 4]
```diff
@@ -342,6 +369,397 @@ void assertTransactionSenderAndValue() {
...
+				sender, receiverAddress, gasLimit, 1234L, Bytes.EMPTY, consensusTime,
+				BigInteger.valueOf(offeredGasPrice).multiply(WEIBARS_TO_TINYBARS),
+				relayer, 10 * ONE_HBAR);
+
+		assertTrue(result.isSuccessful());
+		assertEquals(result.getGasUsed(), gasLimit);
+		assertEquals(receiver.getId().asGrpcContract(), result.toGrpc().getContractID());
+		verify(mutableRelayerAccount,
+				never()).incrementBalance(any());
+		verify(mutableSenderAccount, never()).incrementBalance(any());
```
Comment (Nana-EC) [inline hedera-node/src/test/java/com/hedera/services/contracts/execution/CallEvmTxProcessorTest.java:409]: Missing a `verify(mutableRelayerAccount).decrementBalance(...))` and a `verify(mutableSenderAccount).decrementBalance(...))`

Comment (Nana-EC) [reply hedera-node/src/test/java/com/hedera/services/contracts/execution/CallEvmTxProcessorTest.java:409]: Same for `assertSuccessEthereumTransactionExecutionChargesBothSenderAndRelayerAndRefunds` and `assertSuccessEthereumTransactionExecutionChargesBothSenderAndRelayerAndRefundsOnlyRelayer`

File: test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/contract/HapiEthereumCall.java [hunk 5]
```diff
@@ -267,12 +274,13 @@ protected Consumer<TransactionBody.Builder> opBodyDef(HapiApiSpec spec) throws T
         }
 
         final var longTuple = TupleType.parse("(int64)");
-        final var gasPriceBytes = Bytes.wrap(longTuple.encode(Tuple.of(gasPrice)).array()).toArray();
+        final var gasPriceBytes = Bytes.wrap(longTuple.encode(Tuple.of(gasPrice.longValueExact())).array()).toArray();
```
Comment (Nana-EC) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/contract/HapiEthereumCall.java:277]: can move `longTuple` up to a class field and utilize with something like 
```suggestion
        final var gasPriceBytes = gasLongToBytes(gasPrice);
```

where 
```java
        private byte[] gasLongToBytes(BigInteger gas) {
		return Bytes.wrap(longTuple.encode(Tuple.of(gas.longValueExact())).array()).toArray();
	}
```

File: test-clients/src/main/java/com/hedera/services/bdd/suites/ethereum/HelloWorldEthereumSuite.java [hunk 6]
```diff
@@ -49,7 +49,7 @@ public static void main(String... args) {
     @Override
     public List<HapiApiSpec> getSpecsInSuite() {
         return allOf(
-                ethereumCalls(),
+//                ethereumCalls(),
```
Comment (Nana-EC) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/ethereum/HelloWorldEthereumSuite.java:52]: remove or uncomment

File: test-clients/src/main/java/com/hedera/services/bdd/suites/ethereum/HelloWorldEthereumSuite.java [hunk 7]
```diff
@@ -62,9 +62,9 @@ List<HapiApiSpec> ethereumCalls() {
 
     List<HapiApiSpec> ethereumCreates() {
         return List.of(
-                smallContractCreate(),
-                bigContractCreate(),
-                contractCreateWithConstructorArgs()
+//                smallContractCreate(),
```
Comment (Nana-EC) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/ethereum/HelloWorldEthereumSuite.java:65]: remove or restore

**Review by Nana-EC — COMMENTED** [review]
A few suggestions

**Pull request merged**